### PR TITLE
Support TX rollbacks

### PR DESCRIPTION
**Pull request by ureeves**
This PR is intended to address the transaction rollbacks problem described in #263.

The change entails keeping two states in memory - the `origin` and the `head` state. The caller can choose to perform queries or transactions on either, and the result of transactions will is kept in the `head` state.

When the user deems that enough transactions have been performed on the state they can call `commit` to copy the `head` into the origin.

If the users finds that for some reason they want to discard the state, they can call `reset` to copy `origin` into `head`.

### CODE REVIEW

File: src/state.rs [hunk 1]
```diff
@@ -201,46 +239,24 @@ impl NetworkState {
...
-        *self = fork;
+        // updates made in the forked state. The forked snapshot is set as
+        // the latest, and the latest is hashed and put in the map of
+        // snapshots.
+        mem::swap(&mut self.latest, &mut snap);
+
+        let snap_id = Store::hash(&snap.encode_to_vec());
+        self.snapshots
+            .insert(snap_id, snap)
+            .map_err(VMError::from_store_error)?;
```
Comment (miloszm) [inline src/state.rs:251]: I guess similar logic to the above will be added in ops/transact for inter-contract calls

File: src/state.rs [hunk 2]
```diff
@@ -167,15 +204,16 @@ impl NetworkState {
...
     {
-        let mut context = CallContext::new(self);
+        let mut context =
+            CallContext::new(&mut self.latest, self.modules.clone());
 
         let result =
             context.query(target, Query::from_canon(&query), gas_meter)?;
 
         result.cast().map_err(VMError::from_store_error)
     }
```
Comment (miloszm) [inline src/state.rs:214]: I am not sure we have a mechanism in place which makes it impossible for a query to change state - I guess not, and for this change I think we probably have to make sure that it is not allowed.

File: src/modules.rs [hunk 3]
```diff
@@ -4,13 +4,69 @@
...
+    }
+
+    /// Get a reference to a particular module from the store.
+    pub fn get_module_ref<'a>(&'a self, id: &'a ContractId) -> ModuleRef<'a> {
+        ModuleRef {
+            map_ref: self.0.borrow(),
+            id,
+        }
+    }
+}
```
Comment (miloszm) [inline src/modules.rs:68]: I like this refactoring, before it was mixed up with registration logic

File: src/state.rs (hunk 1 above)
Comment (ureeves) [reply src/state.rs:251]: Iḿ not sure I understand. Could you elaborate?

File: src/modules.rs (hunk 3 above)
Comment (ureeves) [reply src/modules.rs:68]: I thought I would take this as an opportunity. It's related to these changes only tangentially, but in my opinion it's ok.

I'm  not entirely happy with `ModuleRef` as a concept, so if you find a way to remove it and return return a `Option<&BoxedHostModule>` directly from modules I'm all ears.

File: src/state.rs (hunk 2 above)
Comment (ureeves) [reply src/state.rs:214]: Yeah, we don't have such a mechanism in place. An easy solution would be to clone `latest` and discard it, but I would be happier with a solution that takes `&self` instead of `&mut self.

Comment (ureeves) [reply src/state.rs:214]: Fixed this by cloning and making it clear that a `query` doesn't change the state by taking a `&self` instead of `&mut self`. Currently this clones a snapshot - would be nice to not have to do that.

File: src/state.rs (hunk 1 above)
Comment (miloszm) [reply src/state.rs:251]: there are 2 places from which we are calling transact - from state and from ops/transact. You took care of one of these places, here in state.rs, but there is one more place that changes network/contract state, see ops/transact.rs

Comment (miloszm) [reply src/state.rs:253]: as discussed on Friday - the above is not valid if we assume that "transact" is just a name, and the actual meaning is more like Ethereum CALL or DELEGATECALL - in such case it is not needed to add snapshot logic in ops/transact - the name of the method is misleading though

File: src/state.rs [hunk 4]
```diff
@@ -167,30 +204,33 @@ impl NetworkState {
...
+    /// Transact with the contract at `target` address in the latest snapshot,
+    /// returning both the result of the transaction and the snapshot ID of the
+    /// state before applying the transaction.
     pub fn transact<A, R>(
         &mut self,
         target: ContractId,
         transaction: A,
         gas_meter: &mut GasMeter,
-    ) -> Result<R, VMError>
+    ) -> Result<(R, [u8; 32]), VMError>
```
Comment (vlopes11) [inline src/state.rs:224]: This way, it looks like we are generating snapshots per transaction, when it should be per block.

The proposed API binds the future usage of this function to that behavior.

Even if internally we store the snapshots per transaction, this shouldn't be reflected in the API

File: src/state.rs [hunk 5]
```diff
@@ -201,46 +241,37 @@ impl NetworkState {
...
-            .insert(module.module_id(), Box::new(module));
+    /// Reverts the latest state to the given `snap_id`.
+    pub fn revert(&mut self, snap_id: &[u8; 32]) -> Result<(), VMError> {
+        let snap = self
+            .snapshots
+            .remove(snap_id)
+            .map_err(VMError::from_store_error)
+            .transpose()
+            .unwrap_or(Err(VMError::UnknownSnapshot))?;
+
```
Comment (miloszm) [inline src/state.rs:265]: Do we have logic that removes all snapshots more recent than the reverted snapshot? Here we only make the reverted-to snapshot to be the latest, but what about all subsequent snapshots that should be removed?

File: src/state.rs (hunk 4 above)
Comment (miloszm) [reply src/state.rs:224]: Based on what should a rollback be done? I agree that if snapshot is internal to rusk-vm, then we should not expose it.

**Review by miloszm — CHANGES_REQUESTED** [review]
The change looks really good but I wonder about tests.
I'd add at least a single test demonstrating the main rollback scenario.

**Comment by ureeves** [comment]
> The change looks really good but I wonder about tests. I'd add at least a single test demonstrating the main rollback scenario.

You're absolutely right, of course. I'll add a test for `reset` and `commit`.

**Review by miloszm — APPROVED** [review]

File: src/call_context.rs [hunk 6]
```diff
@@ -107,8 +107,7 @@ impl<'a> CallContext<'a> {
 
         let instance: Instance;
 
-        if let Some(module) = self.state.modules().get_module_ref(&target).get()
-        {
+        if let Some(module) = self.state.get_module_ref(&target).get() {
```
Comment (ZER0) [inline src/call_context.rs:110]: I don't think this is part of the scope of this PR, it's not a changed needed.

File: src/state.rs [hunk 7]
```diff
@@ -52,79 +123,78 @@ impl NetworkState {
...
+        self.head.get_contract_mut(contract_id)
+    }
+
+    /// Get a reference to a particular module from the store.
+    pub fn get_module_ref<'a>(
+        &'a self,
+        id: &'a ContractId,
+    ) -> HostModuleRef<'a> {
+        self.modules.get_module_ref(id)
+    }
```
Comment (ZER0) [inline src/state.rs:155]: This is not related to the rollbacks' change.

**Review by ZER0 — CHANGES_REQUESTED** [review]
I would remove the change related to the module, that doesn't seems be related to this PR, plus I will add to `Contracts` structure the `root` method (that returns the root of the head). If @krl still didn't have that implemented properly in `hamt`, we can hash the structure and add a huge `TODO:` to improve it, with the reference of the issue related to that.
But at least we can get the state root for rusk and the node.

File: src/call_context.rs (hunk 6 above)
Comment (ureeves) [reply src/call_context.rs]: Resolved

File: src/state.rs (hunk 7 above)
Comment (ureeves) [reply src/state.rs]: Ditto

**Comment by ureeves** [comment]
> I would remove the change related to the module, that doesn't seems be related to this PR, plus I will add to `Contracts` structure the `root` method (that returns the root of the head). If @krl still didn't have that implemented properly in `hamt`, we can hash the structure and add a huge `TODO:` to improve it, with the reference of the issue related to that. But at least we can get the state root for rusk and the node.

Resolved
